aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -110,55 +141,149 @@ impl<V: TransactionWrite> DataRead<V> {
-            (DataRead::Resolved(_), ReadKind::Exists) => DataRead::Exists(true),
-            (DataRead::Metadata(maybe_metadata), ReadKind::Exists) => {
-                DataRead::Exists(maybe_metadata.is_some())
+            DataRead::Metadata(metadata) => Self::metadata_convert_to(metadata, kind),
+            DataRead::ResourceSize(size) => Self::resource_size_convert_to(*size, kind),
+            DataRead::Exists(exists) => Self::exists_convert_to(*exists, kind),
+        }
+    }
+
+    fn versioned_convert_to(
+        version: &Version,
+        v: &Arc<V>,
+        layout: &Option<Arc<MoveTypeLayout>>,
+        kind: &ReadKind,
+    ) -> Option<DataRead<V>> {
+        match kind {
+            ReadKind::Value => Some(DataRead::Versioned(
+                version.clone(),
+                v.clone(),
+                layout.clone(),
+            )),
+            ReadKind::MetadataAndResourceSize => Some(DataRead::MetadataAndResourceSize(
+                v.as_state_value_metadata(),
+                Self::v_size(v),
+            )),
+            ReadKind::Metadata => Some(DataRead::Metadata(v.as_state_value_metadata())),
+            ReadKind::ResourceSize => Some(DataRead::ResourceSize(Self::v_size(v))),
+            ReadKind::Exists => Some(DataRead::Exists(!v.is_deletion())),
+        }
+    }
+
+    fn resolved_convert_to(v: u128, kind: &ReadKind) -> Option<DataRead<V>> {
+        match kind {
+            ReadKind::Value => Some(DataRead::Resolved(v)),
+            ReadKind::MetadataAndResourceSize => Some(DataRead::MetadataAndResourceSize(
+                Some(StateValueMetadata::none()),
+                Some(serialize(&v).len() as u64),
+            )),
+            ReadKind::Metadata => Some(DataRead::Metadata(Some(StateValueMetadata::none()))),
+            ReadKind::ResourceSize => {
+                Some(DataRead::ResourceSize(Some(serialize(&v).len() as u64)))
             },
-            (_, _) => unreachable!("{:?}, {:?} must be covered", self_kind, kind),
-        })
+            ReadKind::Exists => Some(DataRead::Exists(true)),
+        }
+    }
+
+    fn metadata_and_size_convert_to(
+        maybe_metadata: &Option<StateValueMetadata>,
+        maybe_size: Option<u64>,
+        kind: &ReadKind,
+    ) -> Option<DataRead<V>> {
+        match kind {
+            ReadKind::Value => None,
+            ReadKind::MetadataAndResourceSize => Some(DataRead::MetadataAndResourceSize(
+                maybe_metadata.clone(),
+                maybe_size,
+            )),
+            ReadKind::Metadata => Some(DataRead::Metadata(maybe_metadata.clone())),
+            ReadKind::ResourceSize => Some(DataRead::ResourceSize(maybe_size)),
+            ReadKind::Exists => Some(DataRead::Exists(maybe_metadata.is_some())),
+        }
+    }
+
+    fn metadata_convert_to(
+        maybe_metadata: &Option<StateValueMetadata>,
+        kind: &ReadKind,
+    ) -> Option<DataRead<V>> {
+        match kind {
+            ReadKind::Value => None,
```

Comment:
nit: her and elsewhere you can `ReadKind::Value | ReadKind::MetadataAndResourceSize |  ReadKind::ResourceSize => None,`

Reply:
done